Let jails ip4.addr be added to pf jails table if it isn't local to any interface

kimATwork opened the pull request:
#841
I added the same check that is present in stop.sh, but could just as well be:
`if [ $(bastille config "${_jail}" get interface) == ${bastille_network_loopback} ]; then`

tschettervictor commented:
I don't follow this.

Isn't it necessary to add the jail IP to the firewall table for any interface it is on?

I currently have jails that I connect to the host interface (vtnet0) and they run fine.


Could you give a bit more context.

kimATwork commented:
> Isn't it necessary to add the jail IP to the firewall table for any interface it is on?
A firewall is not required if the jail uses a shared interface (as in 'ifconfig alias'), since NAT is not needed. In that case the jail has its own IP on the LAN. 
> I currently have jails that I connect to the host interface (vtnet0) and they run fine.
I think VNET is one of three ways of setting up to network according to the docs. If ypu try to combine the two others (cloned interface and alias), you'll hit this bug.
> Could you give a bit more context.

kimATwork commented:
> Isn't it necessary to add the jail IP to the firewall table for any interface it is on?

A firewall is not required if the jail uses a shared interface (as in 'ifconfig alias'), since NAT is not needed. In that case the jail has its own ip on the same interface/LAN as the host.

> I currently have jails that I connect to the host interface (vtnet0) and they run fine.

Yes, I do to, but I've never run jails on host interface and cloned interface (bastille0) on the same host at the same time, until now.

> Could you give a bit more context.

Sure.

I have a host, bastille01, with ip 192.168.0.10 on interface xl1. After installing bastille I can create a jail like this:
```
[root@bastille01 ~]# bastille create jail_A 14.2-RELEASE 192.168.0.20 xl1
jail_A: created
```
I can reach jail_A by setting up sshd in the jail and ssh to it from any host in the network.

Then i run out of ip adresses so I configure bastille to use a cloned interface (bastille0), activate pf, and sets up the pf.conf to handle NAT and rdr. As the documentation describes.

Now I can create a jail like this:
```
[root@bastille01 ~]# bastille create jail_B 14.2-RELEASE 172.16.2.3 bastille0
jail_B: created
```
To be able to reach sshd in jail_B from the rest of the network, i need to configure port forwarding:

`[root@bastille01 ~]# bastille rdr host_B tcp 22000 22`


Now I can reach both my jails by ssh. When i check the pf table and bastille it looks like this:
```
[root@bastille01 ~]# pfctl -t jails -T show
   172.16.2.3

[root@bastille01 ~]# bastille list -a
 JID  State  IP Address    Published Ports  Hostname  Release           Path
 57   Up     192.168.0.20  -                jail_A    14.2-RELEASE      /usr/local/bastille/jails/jail_B/root
 65   Up     172.16.2.3    tcp/2200:22      jail_B    14.2-RELEASE      /usr/local/bastille/jails/jail_A/root
```
Then, when i restart jail_A, I can no longer reach it via ssh. What happend? If I add port forwarding, it works again, but that should not be necessary. If I check the pf table again, it now look like this:
```
[root@bastille01 ~]# pfctl -t jails -T show
   172.16.2.3
   192.168.0.20
```
Jail_A (192.168.0.20) should not even be there. So I remove it:
`[root@bastille01 ~]# pfctl -q -t jails -T delete 192.168.0.20`

Now, its reachable again.

Then i restart jail_A and it turns unreachable again. And its ip is back in the pf table again.

 OK, so i'll stop both  jails:
```
   [root@bastille01 /usr/local/share/bastille]# bastille stop jail_A
   nat cleared
   [jail_A]:
   jail_A: removed
   [root@bastille01 /usr/local/share/bastille]# bastille stop jail_B
   [jail_B]:
   jail_B: removed
```
Now, jail_Bs ip is gone from the pf table, but jail_As ip is still present:
```
[root@bastille01 ~]# pfctl -t jails -T show
   192.168.0.20
```
So this means that jail_B, thats using the cloned interface, is working as expected but jail_A only works by adding port redirection (which should not be necessary) or by removing Jail_As ip from pf manually after its started. Also, jail_As ip is still present in the pf table even after its stopped.

The code handling pf in start.sh:

```
  83         ## warn if matching configured (but not online) ip4.addr, ignore if there's no ip4.addr entry
  84         _ip4=$(bastille config "${_jail}" get ip4.addr)
  85         if [ "${_ip4}" != "not set" ]; then
  86             if ifconfig | grep -wF "${_ip4}" >/dev/null; then
  87                 error_notify "Error: IP address (${_ip4}) already in use."
  88                 continue
  89             fi
  90             ## add ip4.addr to firewall table
  91             pfctl -q -t "${bastille_network_pf_table}" -T add "${_ip4}"
  92         fi
```

The code handling pf in stop.sh:

```
 77         ## remove (captured above) ip4.addr from firewall table
 78         if [ -n "${bastille_network_loopback}" ] && [ "${_ip4}" != "not set" ]; then
 79             if grep -qw "interface.*=.*${bastille_network_loopback}" "${bastille_jailsdir}/${_jail}/jail.conf"; then
 80                 pfctl -q -t "${bastille_network_pf_table}" -T delete "${_ip4}"
 81             fi
 82         fi
```

stop.sh line 78 checks if bastille_network_loopback is configured and on line 79 it checks that the jails interface matches the loop back interface, before removing the ip from the firewall.

That check is not done in start.sh, which explains why jail_A is put into the pf table, and never removed.
The root problem is that start.sh should do the same test as stop.sh, keeping jail_A out of the pf table.

So that is what my fix aims to do. I've tested it on my own host.

tschettervictor commented:
So, you're saying that using the alias mode should not have the IP being added to the pf table?
This does make sense, and I'm not sure why the check would exist inside stop.sh and not start.sh
Probably an oversight.

@yaazkal @bmac2

tschettervictor commented:
Might be slight overkill here.

We should not check for the "loopback address" because it is always filled by default with "bastille0"

Also, the "pf_table" is always defined as jails so no point in checking that either.

We only need to check if the "ip4.addr" is "not set" and grep for the "gateway" when running the "route" command.

kimATwork commented:
Really? If i wanted to run my own firewall, for example ipfw, i'd just remove the pf related setup from bastille.conf. But then again, without the tests  it will be more visible if the user types in the wrong IP triggering the pf insert,  without having pf installed. I have committed the changes now.

tschettervictor commented:
There is an open issue requesting support for IPFW.

We are probably going to merge #830 sometime soon, and I've purposely structured it so we can do that.

The issue we have now is

1. How do we tell which is being used? IPFW or PF? (Perhaps another switch inside the conf file? "bastille_network_firewall")

2. How do we tell if the IP is in the local net so as not to add it to the table? (This I think you solve with the route command)

tschettervictor commented:
Your latest changes look good.

I still struggle to replicate.

Couple questions.

1. Why does the jail work the first time it starts? The IP has already been added to the table, so it doesn't make sense that it would work.

2. How are you testing the outbound connection?

PR #830 which you retested now adds and removes the IPs properly for each jail that is non-VNET, without any extra checks. If you still have issues after that, then we can consider merging this one.

kimATwork commented:
> 1. How do we tell which is being used? IPFW or PF? (Perhaps another switch inside the conf file? "bastille_network_firewall")

I haven't user IPFW to handle NAT, only stateful filtering and traffic shaping. Pf and IPFW works fine together in my setup. IPFW as a firewall and pf handling NAT. But i think supporting IPFW would be nice. So using "bastille_network_firewall" default set to "No" seems like a good starting point to me. The user will need to configure the firewall of choice anyway.

> 2. How do we tell if the IP is in the local net so as not to add it to the table? (This I think you solve with the route command)

Yes, I think the route command solves it as far as I can see. It should know the hosts network and should always available.

> Why does the jail work the first time it starts? The IP has already been added to the table, so it doesn't make sense that it would work.

It worked initially because I at that time hadn't set up pf. I ran out of IP addresses, and wanted to solve that by using Bastilles 'loopback' option. After configuring pf etc. the second jail worked as expected using 'bastille0'. Then when i restarted the first jail again I could'n ssh into it even if the second jail using loopback interface worked fine. The rest is history i guess :)

> How are you testing the outbound connection?

Fetching index.html from a known Norwegian news paper.

tschettervictor commented:
Since we are adding this I have added this fix to #830

It should be merged soon.

tschettervictor commented:
This fix was merged in #830 so closing this.
